pkg/common/cns-lib/volume/util.go

```diff
@@ -41,36 +42,92 @@ func validateManager(ctx context.Context, m *defaultManager) error {
 
 // IsDiskAttached checks if the volume is attached to the VM.
 // If the volume is attached to the VM, return disk uuid of the volume, else return empty string
-func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string) (string, error) {
+func IsDiskAttached(ctx context.Context, vm *cnsvsphere.VirtualMachine, volumeID string, checkNVMeController bool) (string, error) {
 	log := logger.GetLogger(ctx)
 	// Verify if the volume id is on the VM backing virtual disk devices
 	vmDevices, err := vm.Device(ctx)
 	if err != nil {
 		log.Errorf("failed to get devices from vm: %s", vm.InventoryPath)
 		return "", err
 	}
+	nvmeControllerKeyToNameMap := make(map[int32]string)
+	for _, device := range vmDevices {
+		if vmDevices.TypeName(device) == "VirtualNVMEController" {
+			var controllerName string
+			if device.GetVirtualDevice().DeviceInfo.GetDescription() != nil {
+				controllerName = device.GetVirtualDevice().DeviceInfo.GetDescription().Label
+			}
+			nvmeControllerKeyToNameMap[device.GetVirtualDevice().Key] = controllerName
+		}
+	}
+
 	for _, device := range vmDevices {
 		if vmDevices.TypeName(device) == "VirtualDisk" {
-			if virtualDisk, ok := device.(*vimtypes.VirtualDisk); ok {
+			if virtualDisk, ok := device.(*types.VirtualDisk); ok {
 				if virtualDisk.VDiskId != nil && virtualDisk.VDiskId.Id == volumeID {
 					virtualDevice := device.GetVirtualDevice()
-					if backing, ok := virtualDevice.Backing.(*vimtypes.VirtualDiskFlatVer2BackingInfo); ok {
+					if value, ok := nvmeControllerKeyToNameMap[virtualDevice.ControllerKey]; ok {
+						log.Info("Found that the disk in contention is attached to NVMe controller")
+						if strings.Contains(value, "NVME") {
+							log.Info("Fetching the backing uuid for the disk on NVMe controller")
+							if backing, ok := virtualDevice.Backing.(*types.VirtualDiskFlatVer2BackingInfo); ok {
+								uuid, err := getNvmeUUID(ctx, backing.Uuid)
+								if err != nil {
+									log.Errorf("failed to convert uuid to  NvmeV13UUID for the vm: %s", vm.InventoryPath)
+									return "", err
+								}
+								log.Infof("Successfully converted diskUUID %s to NvmeV13UUID %s for volume %s on vm %+v", backing.Uuid, uuid, volumeID, vm)
+								return uuid, nil
+							}
+						}
+					}
+					log.Info("Disk in contention is not attached to NVMe controller")
+					if backing, ok := virtualDevice.Backing.(*types.VirtualDiskFlatVer2BackingInfo); ok {
 						log.Infof("Found diskUUID %s for volume %s on vm %+v", backing.Uuid, volumeID, vm)
```

Comment:
Can be moved to debug log.

Reply:
I would let this log be Infof, since I made sure all other logs are debug. This is the only log which will help us debug in prod to verify if disk is attached to scsi or nvme controller